pkg/jobs/hive/hive.go

```diff
@@ -181,3 +187,189 @@ func monitorDeployStatus(client clientv1.Client, hiveset hiveclient.Interface, c
 	}
 	return nil
 }
+
+func UpgradeCluster(client clientv1.Client, clusterName string, curator *clustercuratorv1.ClusterCurator) error {
+	klog.V(0).Info("* Initiate Upgrade")
+	klog.V(2).Info("Looking up managedclusterinfo " + clusterName)
+
+	desiredUpdate := curator.Spec.Upgrade.DesiredUpdate
+
+	managedClusterInfo := managedclusterinfov1beta1.ManagedClusterInfo{}
+	if err := client.Get(context.TODO(), types.NamespacedName{
+		Namespace: clusterName,
+		Name:      clusterName,
+	}, &managedClusterInfo); err != nil {
+		return err
+	}
+
+	klog.V(2).Info("kubevendor ", managedClusterInfo.Status.KubeVendor)
+
+	if managedClusterInfo.Status.KubeVendor != "OpenShift" {
+		return errors.New("Can not upgrade non openshift cluster")
+	}
+
+	if desiredUpdate == "" {
+		return errors.New("Provide valid upgrade version")
+	}
+
+	isValidVersion := false
+	if managedClusterInfo.Status.DistributionInfo.OCP.AvailableUpdates != nil {
+		for _, version := range managedClusterInfo.Status.DistributionInfo.OCP.AvailableUpdates {
+			if version == desiredUpdate {
+				isValidVersion = true
+			}
+		}
+	}
+	if !isValidVersion {
+		return errors.New("Provided version is not valid")
+	}
+
+	klog.V(2).Info("Create managedclusterview " + clusterName)
+
+	managedclusterview := &managedclusterviewv1beta1.ManagedClusterView{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      clusterName,
+			Namespace: clusterName,
+		},
+		Spec: managedclusterviewv1beta1.ViewSpec{
+			Scope: managedclusterviewv1beta1.ViewScope{
+				Group:     "config.openshift.io",
+				Kind:      "ClusterVersion",
+				Name:      "version",
+				Namespace: "",
+				Version:   "v1",
+			},
+		},
+	}
+
+	if err := client.Create(context.TODO(), managedclusterview); err != nil {
+		return err
+	}
+	resultmcview := managedclusterviewv1beta1.ManagedClusterView{}
+	for i := 1; i <= 5; i++ {
+		time.Sleep(utils.PauseFiveSeconds)
+		if err := client.Get(context.TODO(), types.NamespacedName{
+			Namespace: clusterName,
+			Name:      clusterName,
+		}, &resultmcview); err != nil {
+			return err
+		}
+		if resultmcview.Status.Result.Raw != nil {
+			break
+		}
+	}
+
+	resultClusterVersion := resultmcview.Status.Result
+
+	clusterVersion := map[string]interface{}{}
+
+	if resultClusterVersion.Raw != nil {
+		err := json.Unmarshal(resultClusterVersion.Raw, &clusterVersion)
+		utils.CheckError(err)
+	}
+
+	var desiredVersion interface{}
+	if status, ok := clusterVersion["status"]; ok {
+		cvstatus := status.(map[string]interface{})
+		if availableUpdates, ok := cvstatus["availableUpdates"]; ok {
+			if cvAvailableUpdates, ok := availableUpdates.([]interface{}); ok {
+				for _, version := range cvAvailableUpdates {
+					if version.(map[string]interface{})["version"] == desiredUpdate {
+						desiredVersion = version
```

Comment:
these variables are confusing 😂  `disiredUpdate` vs `desiredVersion`, and they are in different type? I'm actually good with the current naming. Just want to mention it can be a nice to have to change to something more easily to tell their difference